src/core/agent/services/connectionService.ts

```diff
@@ -299,21 +306,52 @@ class ConnectionService extends AgentService {
   async getConnectionHistoryById(
     connectionId: string
   ): Promise<ConnectionHistoryItem[]> {
-    let histories: ConnectionHistoryItem[] = [];
     const credentialRecords =
       await this.credentialStorage.getCredentialMetadataByConnectionId(
         connectionId
       );
-    histories = histories.concat(
-      credentialRecords.map((record) => {
-        return {
-          type: ConnectionHistoryType.CREDENTIAL_ACCEPTED,
-          timestamp: record.createdAt.toISOString(),
-          credentialType: record.credentialType,
+    const credentialStatistics: {
+      [key: string]: { name: string; count: number };
+    } = {};
+    credentialRecords.forEach((record) => {
+      if (!credentialStatistics[record.id.replace("metadata:", "")]) {
+        credentialStatistics[record.id.replace("metadata:", "")] = {
+          name: record.credentialType,
+          count: 0,
         };
-      })
+      }
+    });
+
+    const linkedIpexMessages = await this.getLinkedIPEXMessageByConnectionId(
+      connectionId
     );
-    return histories;
+
+    const requestMessages = linkedIpexMessages
+      .sort((a, b) => a.createdAt.getTime() - b.createdAt.getTime())
+      .map((messageRecord) => {
+        const credentialStatistic =
+          credentialStatistics[messageRecord.content.exn.e.acdc.d];
+        let historyType: ConnectionHistoryType;
+        const route = messageRecord.content.exn.r;
+        if (
+          route === ExchangeRoute.IpexGrant &&
+          credentialStatistic.count === 0
+        ) {
+          historyType = ConnectionHistoryType.CREDENTIAL_ISSUANCE;
```

Comment:
I see what you're doing here - but I think it would be better if we stored this extra metadata when storing the IPEX message. We will already have login in the Signify notification service to determine if it's an issuance, update, etc. That way, we don't need to do this calculation every time.